modules/client/src/controllers/ResolveTransferController.ts

```diff
@@ -210,4 +225,81 @@ export class ResolveTransferController extends AbstractController {
+    const { conditionType, paymentId } = params;
+    switch (conditionType) {
+      case ConditionalTransferTypes.HashLockTransfer: {
+        const { preImage } = params as PublicParams.ResolveHashLockTransfer;
+        return !!preImage;
+      }
+      case ConditionalTransferTypes.GraphTransfer: {
+        const { responseCID, signature } = params as PublicParams.ResolveGraphTransfer;
+        return !!responseCID && !!signature;
+      }
+      case ConditionalTransferTypes.SignedTransfer: {
+        const { data, signature } = params as PublicParams.ResolveSignedTransfer;
+        return !!data && !!signature;
+      }
+      case ConditionalTransferTypes.LinkedTransfer: {
+        const { preImage } = params as PublicParams.ResolveLinkedTransfer;
+        return !!preImage;
+      }
+      default: {
+        const c: never = conditionType;
+        this.log.error(`[${paymentId}] Unsupported conditionType ${c}`);
+      }
+    }
+    throw new Error(`Invalid condition type: ${conditionType}`);
+  }
+
+  private async handleCancellation(
+    params: PublicParams.ResolveCondition,
+  ): Promise<PublicResults.ResolveCondition> {
+    const { conditionType, paymentId } = params;
+    const appDefinition = this.connext.appRegistry.find((app) => app.name === conditionType)
+      .appDefinitionAddress;
+    const apps = await this.connext.getAppInstances();
+    const paymentApp = apps.find((app) => {
+      const participants = (app.latestState as GenericConditionalTransferAppState).coinTransfers.map(
+        (t) => t.to,
+      );
+      return (
+        app.appDefinition === appDefinition &&
+        app.meta.paymentId === paymentId &&
+        participants.includes(this.connext.signerAddress)
+      );
+    });
+
+    if (!paymentApp) {
+      throw new Error(`Cannot find payment associated with ${paymentId}`);
+    }
+
+    const ret = {
+      appIdentityHash: paymentApp.identityHash,
+      amount: (paymentApp.latestState as GenericConditionalTransferAppState).coinTransfers[0]
+        .amount,
+      assetId: paymentApp.outcomeInterpreterParameters["tokenAddress"],
+      meta: paymentApp.meta,
+      paymentId: params.paymentId,
+      sender: paymentApp.meta.sender,
+    };
+
+    let action: HashLockTransferAppAction | undefined = undefined;
+    // IFF payment app has an expiry, the app has not expired, and the user
+    // is the receiver of the payment, play an invalid action
+    const state = paymentApp.latestState as any;
+    const block = await this.ethProvider.getBlockNumber();
+    if (
+      state.coinTransfers[1].to === this.connext.signerAddress &&
+      toBN(state.expiry || 0).gt(block)
+    ) {
+      action = { preImage: HashZero };
+    }
+    this.log.info(`[${paymentId}] Uninstalling transfer app with action ${stringify(action)}`);
+    await this.connext.uninstallApp(paymentApp.identityHash, action);
+    this.log.info(`[${paymentId}] Uninstalled transfer app ${paymentApp.identityHash}`);
+    return ret;
```

Comment:
This file is unnecessarily complicated. You're already doing `if(state.expiry)...`, so you might as well just do this:

```
if(state.expiry && state.expiry > currentBlockNumber) {
   await uninstallApp(appIdentityHash) // uninstall without taking action
} else {
   await uninstallApp(action, appIdentityHash) // uninstall with action - this will cover *all* other cases!!
```